Allows the name of a collection to not require an underscore immediat…

qchaupds opened the pull request:
…ely after the "collection" token.

Closes NASA-PDS/validate#310

This pull request resolves one issue: A collection file that does not has an underscore after "collection" is now recognized as a collection.
Previously the user received a warning about identifer not belong to any collection is now resolved.

Tested in DEV for #310:

Previously, the user saw a warning and now it does not because the dash after the "collection" token is now recognized

% validate -R pds4.bundle --skip-content-validation -r report_github310_bundle_valid.json -s json -t src/test/resources/github310/valid/bundle.xml

The test below should raise 2 warnings for identifiers not belong to any collection because the name of the collection file is collectioX-data-bst-1.0.xml,
which does not qualify as a collection file.

% validate -R pds4.bundle --skip-content-validation -r report_github310_bundle_invalid.json -s json -t src/test/resources/github310/invalid/bundle.xml

Commit:
Allows the name of a collection to not require an underscore immediately after the "collection" token.

1. Add github310 to src/test/resources directory.
2. Use alternate listFiles() function to avoid infinite loop in BundleManager.java
3. Ease requirement for an underscore immediately after the "collection" token in BundleManager.java
4. Commented out extraneous println statements in FieldValueValidator.java
5. Add function signature for crawl() to use list of file extensions in Crawler.java
6. Add function to craw() to use list of file extensions in FileCrawler.java
7. Consolidate a block of code to refinedFoundList() function so it can be reused in FileCrawler.java
8. Ease requirement for an underscore immediately after the "collection" token in CollectionReferentialIntegrityRule.java
9. Use alternate listFiles() function to avoid infinite loop in CollectionReferentialIntegrityRule.java
10. Ease requirement for an underscore immediately after the "collection" token in FindUnreferencedIdentifiers.java
11. Use alternate listFiles() function to avoid infinite loop in FindUnreferencedIdentifiers.java
12. Add 3 new constants to consolidate the locations of patterns to one place in Constants.java
13. Add two new tests for github310 in feature file validate.feature

Refs:

https://github.com/NASA-PDS/validate/issues/310 Validate missing collections in bundle after CCB-282 updates

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/BundleManager.java:
```diff
@@ -452,7 +471,12 @@ public static Target getLatestBundle() {
+            //allFiles = crawler.crawl(new File(dirName).toURI().toURL(),regexFileFilter);
+            // Note: For some strange reason, the crawler goes into an infinite loop using the above call
+            //       so we will use an alternate call to get the list of files.
+            String[] extensions = new String[1];
```
@qchaupds can we set this as a list constant and convert as needed to an array for this method?

same goes for the "collection" string below on line 479. i could imagine this being rough trying to track down in the future in the event they need to changed for whatever reason.

jordanpadams replied:
@qchaupds also, in the future, can we just make sure we go through and "resolve" each of these comments as you complete them? i would think of the comments as a checklist. this way I know you saw them and made the fix. otherwise, I have to go back and look at my comments and see if it actually changes.

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/BundleManager.java:
```diff
@@ -402,7 +412,14 @@ public static Target getLatestBundle() {
+            // Note: For some strange reason, the crawler goes into an infinite loop using the above call
+            //       so we will use an alternate call to get the list of files.
+            String[] extensions = new String[1];
+            extensions[0] = "xml";  // Note that the extension does not contain the dot.
```
same comment here regarding using a list constant and "bundle" name constant

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/BundleManager.java:
```diff
@@ -267,7 +272,12 @@ public static Target getLatestBundle() {
+                    // Note: For some strange reason, the crawler goes into an infinite loop using the above call
+                    //       so we will use an alternate call to get the list of files.
+                    String[] extensions = new String[1];
+                    extensions[0] = "xml";  // Note that the extension does not contain the dot.
```
same goes here as other comments regarding constants

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/BundleManager.java:
```diff
@@ -188,7 +188,12 @@ public static Target getLatestBundle() {
+                    //kids = crawler.crawl(dir.getUrl(), regexFileFilter); 
+                    // Note: For some strange reason, the crawler goes into an infinite loop using the above call
+                    //       so we will use an alternate call to get the list of files.
+                    String[] extensions = new String[1];
```
same goes here as other comments regarding constants

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/content/table/FieldValueValidator.java:
```diff
@@ -422,7 +422,7 @@ private void checkMinMax(String value, Double minimum, Double maximum,
 
         LOG.debug("checkMinMax:originalValue,value,minimum,maximum {},{},{},{}",originalValue,value,minimum,maximum);
-        System.out.println("checkMinMax:originalValue,value " + originalValue + " " + value);
+        //System.out.println("checkMinMax:originalValue,value " + originalValue + " " + value);
```
can we just remove this code instead of commenting it out? i think we have enough lines of code in this software to remove this :-)

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/crawler/FileCrawler.java:
```diff
@@ -61,33 +117,81 @@ public FileCrawler() {
-    for (File file : FileUtils.listFiles(directory, fileFilter, null)) {
+    LOG.debug("crawl:getDirectories {}",getDirectories);
+
+// Used by developer to print the stack of this function to debug infinite loop.
```
@qchaupds now that the bug has been found, can we remove this dead code?

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/crawler/FileCrawler.java:
```diff
@@ -61,33 +117,81 @@ public FileCrawler() {
+    LOG.debug("crawl:LISTING_FILES:directory,fileFilter,this.fileFilter {},[{}],[{}]",directory,fileFilter,this.fileFilter);
+
+    Collection<File> collections = FileUtils.listFiles(directory, fileFilter, null);
+//if (2 == 3) {
```
@qchaupds same goes with removing this dead code?

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/crawler/FileCrawler.java:
```diff
@@ -61,33 +117,81 @@ public FileCrawler() {
+
+    results = this.refinedFoundList(collections, fileUrl, directory, getDirectories, null);
+
+if (2 == 3) {
```
@qchaupds ? this seems like debug code?

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/CollectionReferentialIntegrityRule.java:
```diff
@@ -77,7 +78,13 @@ public void collectionReferentialIntegrityRule() {
+      //List<Target> children = crawler.crawl(getTarget(), regexFileFilter);
+      // Note: For some strange reason, the crawler goes into an infinite loop using the above call
+      //       so we will use an alternate call to get the list of collection files.
+      String[] extensions = new String[1];
```
@qchaupds same as other comments regarding using constants

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FindUnreferencedIdentifiers.java:
```diff
@@ -101,6 +104,8 @@ public void findUnreferencedIdentifiers() {
                     "Identifier '" + id.toString() + "' is not a member "
                     +"of any " + memberType + " within the given target"), 
                 locationUrl));
+            //String errorMessage = "Identifier '" + id.toString() + "' is not a member " +"of any " + memberType + " within the given target";
```
needed? or can be removed?

jordanpadams commented on src/test/resources/github310/invalid/data/MAGBST051782340_V1.TAB:
```diff
@@ -0,0 +1,9600 @@
+   28402312.00,    43,   108,   432
```
@qchaupds unless we actually need to do content validation, can we make sure we do not include these large data files in the repo? i would prefer these be empty files and we use the `--skip-content-validation` flag. that goes for all our tests for that matter. our repo will start to get massive with all these test data products if we aren't careful.

jordanpadams commented on src/test/resources/github310/invalid/data/MAGBST051790239_V1.TAB:
```diff
@@ -0,0 +1,9600 @@
+   28413019.60,   104,   169,   376
```
@qchaupds same goes for this test data product

jordanpadams commented on src/test/resources/github310/valid/data/MAGBST051782340_V1.TAB:
```diff
@@ -0,0 +1,9600 @@
+   28402312.00,    43,   108,   432
```
@qchaupds same goes for this test data product

jordanpadams commented on src/test/resources/github310/valid/data/MAGBST051790239_V1.TAB:
```diff
@@ -0,0 +1,9600 @@
+   28413019.60,   104,   169,   376
```
@qchaupds same goes for this test data product

Commit:
PR Response 1: Make use of defined constants in one place to avoid hunting in many places

1.  Use defined constants in and remove un-needed comments in BundleManager.java
2.  Remove un-needed comments in FieldValueValidator.java
3.  Remove un-used codes in FileCrawler.java
4.  Use defined constants in CollectionReferentialIntegrityRule.java
5.  Remove un-needed comments in FindUnreferencedIdentifiers.java
6.  Add defined constants in Constants.java
7.  Reduce test resources to zero size to reduce git repository.  Modify file_size and checksum tags to reflect new parameters.

Refs:

https://github.com/NASA-PDS/validate/issues/310 Validate missing collections in bundle after CCB-282 updates

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/BundleManager.java:
```diff
@@ -46,10 +48,12 @@
-      Pattern.compile("(.*_)*collection(_.*)*\\.xml", Pattern.CASE_INSENSITIVE);
-    private static final Pattern BUNDLE_LABEL_PATTERN =
-      Pattern.compile("(.*_)*bundle(_.*)*\\.xml", Pattern.CASE_INSENSITIVE);
+    private static final Pattern COLLECTION_LABEL_PATTERN = Constants.COLLECTION_LABEL_PATTERN; // Ease the requirement to have an underscore after 'collection'.
```
@qchaupds not a big deal, but is there a particular reason we are setting a new status variable with a Constant set elsewhere? It just seems like someone should just use the `Constants` class to get these values and not set them again within this class.

like I said, not a huge deal, but seems unnecessary?

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/BundleManager.java:
```diff
@@ -448,18 +466,22 @@ public static Target getLatestBundle() {
+            //allFiles = crawler.crawl(new File(dirName).toURI().toURL(),regexFileFilter);
+            // Note: For some strange reason, the crawler goes into an infinite loop using the above call
+            //       so we will use an alternate call to get the list of files.
+            LABEL_EXTENSIONS_LIST[0] = LABEL_EXTENSION;
```
@qchaupds similar to my other comment, this doesn't really seem like good programming practice? I feel like the array should be set as a constant when the constant is declared. not the end of the world, but just something to keep in mind moving forward.
